bob-access/src/credentials.rs

```diff
@@ -77,44 +153,96 @@ impl Credentials {
     }
 }
 
-#[derive(Debug, Default)]
-pub struct CredentialsBuilder {
+#[derive(Default)]
+pub struct RequestCredentialsBuilder {
     kind: Option<CredentialsKind>,
     address: Option<SocketAddr>,
 }
 
-impl CredentialsBuilder {
+impl RequestCredentialsBuilder {
+    pub fn with_username_password(
+        mut self,
+        username: impl Into<String>,
+        password: impl Into<String>,
+    ) -> Self {
+        self.kind = Some(CredentialsKind::Basic {
+            username: username.into(),
+            password: password.into(),
+        });
+        self
+    }
+
+    pub fn with_token(mut self, token: impl Into<String>) -> Self {
+        self.kind = Some(CredentialsKind::Token(token.into()));
+        self
+    }
+
+    pub fn with_address(mut self, address: Option<SocketAddr>) -> Self {
+        self.address = address;
+        self
+    }
+
+    pub fn with_nodename(mut self, node_name: impl Into<String>) -> Self {
+        self.kind = Some(CredentialsKind::InterNode(node_name.into()));
+        self
+    }
+
+    pub fn build(self) -> RequestCredentials {
+        RequestCredentials {
+            address: self.address,
+            kind: self.kind,
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct DeclaredCredentialsBuilder {
+    kind: Option<CredentialsKind>,
+    address: Option<Vec<SocketAddr>>,
+    hostname: Option<String>,
+    resolve_state: ResolveState,
+}
+
+impl DeclaredCredentialsBuilder {
     pub fn with_username_password(
-        &mut self,
+        mut self,
         username: impl Into<String>,
         password: impl Into<String>,
-    ) -> &mut Self {
+    ) -> Self {
         self.kind = Some(CredentialsKind::Basic {
             username: username.into(),
             password: password.into(),
         });
         self
     }
 
-    pub fn with_token(&mut self, token: impl Into<String>) -> &mut Self {
+    pub fn with_token(mut self, token: impl Into<String>) -> Self {
         self.kind = Some(CredentialsKind::Token(token.into()));
         self
     }
 
-    pub fn with_address(&mut self, address: Option<SocketAddr>) -> &mut Self {
+    pub fn with_address(mut self, address: Option<Vec<SocketAddr>>) -> Self {
```

Comment:
`with_addresses`
Also, you can add a separate function `with_address` which receives a single `SocketAddr` and wraps it with vector in its body